### PR TITLE
fix(gatsby-source-contentful): Prevent null pointer exception

### LINKED ISSUE #35216: gatsby-source-contentful Cannot read properties of null (reading 'url')
### Preliminary Checks

- [X] This issue is not a duplicate. Before opening a new issue, please search existing issues: https://github.com/gatsbyjs/gatsby/issues
- [X] This issue is not a question, feature request, RFC, or anything other than a bug report directly related to Gatsby. Please post those things in GitHub Discussions: https://github.com/gatsbyjs/gatsby/discussions

### Description

I've run into another breaking issue with gatsby-source-contentful. It seems to relate to the changes made for the Image CDN beta and likely is related to our locale setup. I've created a reproduction.

The Contentful configuration is simple. Add a second locale. In the test environment we have English and Spanish. Spanish is configured without a fallback. We do this so that we can tell when content isn't translated and most often we hide it.

image

Then, I've just created two image assets with only the default English one uploaded.

From there, just run `npm run develop` and it will crash with:

```
 ERROR #11321  PLUGIN

"gatsby-source-contentful" threw an error while running the sourceNodes lifecycle:

Cannot read properties of null (reading 'url')

  TypeError: Cannot read properties of null (reading 'url')
  
  - normalize.js:651 
    [gatsby-contentful-url]/[gatsby-source-contentful]/normalize.js:651:26
  
  - Array.forEach
  
  - normalize.js:619 createAssetNodes
    [gatsby-contentful-url]/[gatsby-source-contentful]/normalize.js:619:11
  
  - source-nodes.js:305 Object.sourceNodes
    [gatsby-contentful-url]/[gatsby-source-contentful]/source-nodes.js:305:75
  
  - task_queues:96 processTicksAndRejections
    node:internal/process/task_queues:96:5
  
  - api-runner-node.js:462 runAPI
    [gatsby-contentful-url]/[gatsby]/src/utils/api-runner-node.js:462:16
```

### Reproduction Link

https://github.com/afladmark/gatsby-contentful-url

### Steps to Reproduce

1. Download the reproduction repro
2. `npm install` then `npm run develop`
3. See the above error

### Expected Result

Not to error.

### Actual Result

The error above is displayed.

### Environment

```shell
System:
    OS: macOS 12.3
    CPU: (10) arm64 Apple M1 Max
    Shell: 5.8 - /bin/zsh
  Binaries:
    Node: 16.13.2 - ~/.nvm/versions/node/v16.13.2/bin/node
    Yarn: 1.22.17 - ~/.yarn/bin/yarn
    npm: 8.5.3 - ~/.nvm/versions/node/v16.13.2/bin/npm
  Browsers:
    Chrome: 99.0.4844.83
    Edge: 99.0.1150.46
    Firefox: 98.0.1
    Safari: 15.4
  npmPackages:
    gatsby: ^4.10.2 => 4.10.2 
    gatsby-plugin-image: ^2.10.1 => 2.10.1 
    gatsby-plugin-sharp: ^4.10.2 => 4.10.2 
    gatsby-source-contentful: ^7.8.2 => 7.8.2
```

### Config Flags

_No response_

### PR DESCRIPTION
**Pull request by xavivars**
## Description

This MR prevents a null pointer when creating asset nodes that are not configured on some languages.

## Related Issues

Fixes #35216

### CODE REVIEW

**Comment by wardpeet** [comment]
Any chance you can update a test?

**Comment by xavivars** [comment]
@wardpeet , is the solution "good enough"? TBH, I'm not sure creating "broken assets" makes sense.

image

I needed this to (at least) make my build pass. If you think this is good enough, I can try to add a test.

**Comment by axe312ger** [comment]
> @wardpeet , is the solution "good enough"? TBH, I'm not sure creating "broken assets" makes sense.

Yes, I think it does make sense. It is a situation that can be caused within Contentful, when using Preview API:

When you create a new entry, the auto save can be triggered. That will save entities with required fields that are empty.

Entries in current version of the plugin are not affected by this. This is something I have to address in #31385, as the new schema generation will finally mirror the fields from Contentful instead of guessing them.

**Comment by axe312ger** [comment]
As @xavivars gave me write access to his repo, I'll update this PR with a test in a moment :)

**Comment by axe312ger** [comment]
My test repo looks like the following. It should mirror all cases of published/unpublished, valid/invalid, translated/untranslated and has no locale fallback enabled.

I'll add tests that emulate a build process using data from this space. One for Delivery API and one for Preview API.

This should cover a lot of use cases, that are not covered by our E2E tests. (They rely on rendering the Delivery API only)

Screenshot 2022-04-07 at 15 04 56
Screenshot 2022-04-07 at 15 05 05
Screenshot 2022-04-07 at 15 05 13

File: packages/gatsby-source-contentful/src/normalize.js [hunk 1]
```diff
@@ -727,7 +729,7 @@ export const createAssetNodes = ({
       updatedAt: assetItem.sys.updatedAt,
       parent: null,
       children: [],
-      file,
+      file: Object.keys(file).length ? file : null,
```
Comment (wardpeet) [inline packages/gatsby-source-contentful/src/normalize.js:732]: can you expand on this check? Why object.keys

**Review by wardpeet — COMMENTED** [review]
Can't we ignore creating the asset if the file is null? That sounds like a better path forward, no?

**Comment by axe312ger** [comment]
@wardpeet i think you are right. There is no use in importing "broken" assets. If we simply ignore empty assets, the implementation on user/project level should be simplified as well.

I'll change it!

**Comment by axe312ger** [comment]
I pushed the change, assets without any field will now be skipped.

**The good part:**

This is a performance improvement! Less nodes!

**The bad part:**

This could be seen as a breaking change, as data that was existing is suddenly gone.

On the other hand, everyone affected by this potentially breaking change is also affected by the bug and can't build (As long their content has assets with missing files)

**Comment by axe312ger** [comment]
@wardpeet 

The CI now suffers of `Error: MDB_PROBLEM: Unexpected problem - txn should abort` (https://github.com/gatsbyjs/gatsby/issues/35055#issuecomment-1080592765)

https://app.circleci.com/pipelines/github/gatsbyjs/gatsby/81481/workflows/9efc942e-361d-4d52-a5d7-4c48eec52eb9/jobs/965650

File: packages/gatsby-source-contentful/src/normalize.js [hunk 2]
```diff
@@ -740,10 +748,10 @@ export const createAssetNodes = ({
       },
       url: `https:${file.url}`,
       placeholderUrl: `https:${file.url}?w=%width%&h=%height%`,
-      mimeType: file.contentType,
-      filename: file.fileName,
-      width: file.details?.image?.width,
-      height: file.details?.image?.height,
+      mimeType: file.contentType ?? null,
+      filename: file.fileName ?? null,
```
Comment (wardpeet) [inline packages/gatsby-source-contentful/src/normalize.js:752]: These two props can't be null
```suggestion
      mimeType: file.contentType,
      filename: file.fileName,
```

Comment (axe312ger) [reply packages/gatsby-source-contentful/src/normalize.js:754]: @wardpeet Unfortunately they can in some cases when using the Preview API and a (big) asset is still processed by Contentful

I talked to the API team about it. So I added another check for `file.url` to exist and more comments to this section.

Comment (wardpeet) [reply packages/gatsby-source-contentful/src/normalize.js:754]: we can't create an asset when these things are null cause Image CDN needs them.

Comment (axe312ger) [reply packages/gatsby-source-contentful/src/normalize.js:754]: alright, then lets require them as well! :)

**Comment by axe312ger** [comment]
@LekoArts 

The change/commit documentation would be:


> Assets without a file, either empty or untranslated without locale fallback, will no more be available within Gatsby.
>
> This ensures the new Image CDN feature is working properly and simplifies the check for the existence of an (translated) asset.
>
> Fixes #35216


You could argue that this is a breaking change, as it might require changes in the code, especially when handling assets without locale fallbacks. On the other hand, everyone benefitting from this fix, won't be able to build their site right now.

This is likely to prevent multiple projects from upgrading to Gatsby v4.

**Comment by wardpeet** [comment]
@axe312ger can you sketch when it would be breaking change? (Can this happen outside of preview?)

**Comment by axe312ger** [comment]
@wardpeet:

The change requires to change code that handles non-existent assets. Before this, assets could exist, and users might checked if `asset.file.url` did exist. Now the whole asset won't be there anymore.

For users using a space without locale fallback, the page is likely not to build at all as it will fail due to incomplete assets in translation.

The more I think about it, the less it is breaking. It is either working now for you, so you probably don't have to change anything, if its not working for you, this is the patch you need.


I'd say lets publish it to canary/next and see?

**Review by wardpeet — APPROVED** [review]

**Pull request merged**